core/tx_pool_test.go

```diff
@@ -754,6 +761,55 @@ func TestTransactionQueueAccountLimiting(t *testing.T) {
 	}
 }
 
+// Test the limit on transaction size is enforced correctly.
+// This test verifies every transaction having allowed size
+// is added to the pool, and longer transactions are rejected.
+func TestAllowedTxSize(t *testing.T) {
+	t.Parallel()
+
+	// Create a test account and fund it
+	pool, key := setupTxPool()
+	defer pool.Stop()
+
+	account, _ := deriveSender(transaction(0, 0, key))
+	pool.currentState.AddBalance(account, big.NewInt(1024*int64(pool.currentMaxGas)))
+
+	// Compute maximal data size for transactions (lower bound).
+	// It is assumed the fields in the transaction (except of the data) are:
+	// nonce : at most 32 bytes
+	// gasPrice : at most 32 bytes
+	// gasLimit : at most 32 bytes
+	// To address : 20 bytes
+	// value : at most 32 bytes
+	// signature : 65 bytes
+	// All those fields are summed up to at most 213 bytes.
+	txSizeWithoutData := uint64(213)
+	maxDataSize := pool.maxTxSize - txSizeWithoutData
+
+	// Increase block gas limit to infinity
+	pool.currentMaxGas = 2 << 60
+
+	// Try adding a transaction with maximal allowed size
+	if err := pool.validateTx(pricedDataTransaction(0, pool.currentMaxGas, big.NewInt(0), key, maxDataSize), true); err != nil {
+		t.Fatalf("failed to add transaction of size close to maximal: %d, %v", int(pricedDataTransaction(0, pool.currentMaxGas, big.NewInt(1), key, maxDataSize).Size()), err)
+	}
+
+	// Try adding a transaction with random allowed size
+	if err := pool.validateTx(pricedDataTransaction(0, pool.currentMaxGas, big.NewInt(0), key, uint64(rand.Intn(int(maxDataSize)))), true); err != nil {
+		t.Fatalf("failed to add transaction of random allowed size: %v", err)
+	}
+
+	// Try adding a transaction of minimal not allowed size
+	if pool.validateTx(pricedDataTransaction(0, pool.currentMaxGas, big.NewInt(0), key, pool.maxTxSize), true) == nil {
+		t.Fatalf("expected rejection on slightly oversize transaction")
+	}
+
+	// Try adding a transaction of random not allowed size
+	if pool.validateTx(pricedDataTransaction(0, pool.currentMaxGas, big.NewInt(1), key, maxDataSize+1+uint64(rand.Intn(int(10*pool.maxTxSize)))), true) == nil {
+		t.Fatalf("expected rejection on oversize transaction")
+	}
```

Comment:
`pool.validateTx` is an internal method. Please use `addRemoteSync` to test transaction inclusion. That ensures that if we reorganize the internals, the test remains valid and tests what we actually want to test. It's also how all the other tests above work.

Reply:
(Yes, addRemoteSync is also internal, but it's just a helper for AddRemote to not have to explicitly wait for the update loop to run).